rfcs/2022--01-06-8596-tcp-source-backpressure.md

```diff
@@ -0,0 +1,160 @@
+# RFC 8596 - 2022-01-06 - TCP Source Backpressure
+
+Backpressure is necessary in order to limit the total number of events that Vector has to hold when a sink is
+unable to process as quickly as a source accepts new events.
+
+TCP sources today are processed on a per-connection basis. For each connection, a request is decoded, and all the events
+contained in that request are sent to the pipeline. Backpressure is applied when flushing events into the pipeline.
+When this happens, it prevents that one specific connection from decoding another request until there is enough
+room in the pipeline for flushing to complete. It does not, however, prevent new connections from being opened,
+which will decode another request. If the number of connections are unbounded, then the total number of events that
+Vector has to hold in memory is also unbounded, causing excessive memory usage, or crashing Vector entirely.
+
+The goal here is to set an upper bound on the number of events held by Vector by propagating backpressure
+to the source of events.
+
+## Context
+
+- [add config option to limit source tcp connections (off by default)](https://github.com/vectordotdev/vector/pull/10491)
+- [Epic for source backpressure handling in all sources](https://github.com/vectordotdev/vector/issues/8820)
+
+## Cross cutting concerns
+
+## Scope
+
+### In scope
+
+- Limit the number of in-flight events for TCP sources.
+  - fluent
+  - socket (tcp mode only)
+  - logstash
+  - statsd
+  - syslog
```

Comment:
Do any of these protocols support backpressure? If they do -- or if one comes up in the future -- would we simply use that protocol's backpressure mechanism?

Reply:
I'm not aware of any back-pressure mechanism for these protocols. If they do add one in the future, and it adds value beyond what this RFC adds, it would probably make sense to add support for it.